### PR TITLE
Display Swagger UI with docs from app.py (#1163)

### PR DESCRIPTION
**Pull request by jessebrennan**
Things outstanding:
- The docs are very much incomplete. This need to be done in a follow-up ticket.
- We need to decide how to deal with/display API versioning

I have a script that somewhat improves the development process for writing the docs by rendering them, unfortunately I cannot figure out how to display validation. I'm unsure whether or not to include it.

For an example of what this looks like, check out https://service.jesse0.dev.explore.data.humancellatlas.org/

### CODE REVIEW

File: lambdas/service/lambda-policy.json.template.py [hunk 1]
```diff
@@ -122,6 +122,15 @@
...
                 f"arn:aws:states:{aws.region_name}:{aws.account}:execution:{config.cart_export_state_machine_name}"
             ]
+        },
+        {
+            "Effect": "Allow",
+            "Action": [
+                "apigateway:GET"
+            ],
+            "Resource": [
+                f"arn:aws:apigateway:{aws.region_name}::/restapis/{aws.api_gateway_id('service')}/stages/{config.deployment_stage}/exports/oas30"
```
Comment (danielsotirhos) [inline lambdas/service/lambda-policy.json.template.py]: Line longer than 120 char, however I'm not certain the "fix" below makes things better. Possibly?
```suggestion
                f"arn:aws:apigateway:{aws.region_name}::"
                f"/restapis/{aws.api_gateway_id('service')}/stages/{config.deployment_stage}/exports/oas30"
```

Comment (jessebrennan) [reply lambdas/service/lambda-policy.json.template.py]: I'll leave it and see if Hannes complains.

**Comment by jessebrennan** [comment]
@hannes-ucsc I assigned you for review. One of the builds is failing, but I think it's a Travis error; I would appreciate your insight on the issue.

File: src/azul/deployment.py [hunk 2]
```diff
@@ -57,8 +57,18 @@ def secretsmanager(self):
     def dynamo(self, endpoint_url, region_name):
         return boto3.resource('dynamodb', endpoint_url=endpoint_url, region_name=region_name)
 
-    def api_gateway_id(self, function_name: str, validate=True) -> Optional[str]:
+    def api_gateway_export(self, gateway_id):
+        response = self.apigateway.get_export(
+            restApiId=gateway_id,
```
Comment (hannes-ucsc) [inline src/azul/deployment.py]: We're still doing aligned indent for function arguments.

File: lambdas/service/lambda-policy.json.template.py (hunk 1 above)
Comment (hannes-ucsc) [reply lambdas/service/lambda-policy.json.template.py]: I would wrap it there, too.

File: scripts/apidev.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+from http.server import HTTPServer, SimpleHTTPRequestHandler
+
+from watchdog.events import FileSystemEventHandler
+from watchdog.observers import Observer
+
+from azul.deployment import aws
+from azul.modules import load_app_module
+from azul.openapi import annotated_specs
+
+SPEC_FILE = 'openapi.json'
```
Comment (hannes-ucsc) [inline scripts/apidev.py]: We don't do upper case constants.

File: lambdas/service/Makefile [hunk 4]
```diff
@@ -12,6 +12,7 @@ wheels:
 	unzip $(azul_home)/lambdas/.wheels/jsonobject-0.9.9-cp36-cp36m-linux_x86_64.whl -d vendor
 
 deploy: state wheels
+	cp static/openapi.html vendor/
```
[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply lambdas/service/Makefile]: In fact I think the whole `static` directory should be in `vendor/` then we can add more files later without having to add inclusions rules to .gitignore.

File: lambdas/service/app.py [hunk 5]
```diff
@@ -106,11 +119,43 @@ def jwt_auth(auth_request) -> AuthResponse:
...
+@app.route('/openapi', methods=['GET'], cors=True)
+def openapi():
+    gateway_id = os.environ['AZUL_SERVICE_API_GATEWAY_ID']
+    specs = annotated_specs(gateway_id, app, api_info)
+    return Response(status_code=200,
+                    headers={"content-type": "application/json"},
+                    body=specs)
+
+
+@spec(
```
[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply lambdas/service/app.py]: Also, to eliminate one indention level, how about this?

```suggestion
@spec({
```

File: lambdas/service/app.py [hunk 6]
```diff
@@ -106,11 +119,43 @@ def jwt_auth(auth_request) -> AuthResponse:
...
+                'name': 'foo',
+                'description': 'This is not a real parameter',
+                'required': False,
+                'schema': {'$ref': "#/components/schemas/Empty"}
+            }
+        ],
+        'tags': ['Health']
+    }
+)
 @app.route('/health/basic', methods=['GET'], cors=True)
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]: And lets consider this 

```suggestion
@app.route('/health/basic', methods=['GET'], cors=True, openapi={
})
```

File: src/azul/deployment.py [hunk 7]
```diff
@@ -88,10 +98,13 @@ def es_endpoint(self) -> Netloc:
...
     @property
     def lambda_env(self) -> Mapping[str, str]:
-        return config.lambda_env(self.es_endpoint)
+        return {
+            **config.lambda_env(self.es_endpoint),
+            'AZUL_SERVICE_API_GATEWAY_ID': aws.api_gateway_id('service', validate=True)
+        }
 
     def get_lambda_arn(self, function_name, suffix):
-        return f"arn:aws:lambda:{aws.region_name}:{aws.account}:function:{function_name}-{suffix}"
```
Comment (hannes-ucsc) [inline src/azul/deployment.py]: There may be more. Fix them all and move to separate commit if you haven't done so already.

File: src/azul/openapi.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
+import copy
+from typing import MutableMapping, List
+
+from azul.deployment import aws
+from azul.types import JSON
+
+
+def spec(description: JSON):
```
Comment (hannes-ucsc) [inline src/azul/openapi.py]: In the `@app.route(…, openapi=…` approach, our `AzulChalicApp.route` override would call this function. The reason I like tying this to `app.route` is that `@spec` doesn't make sense by itself, it should only be used on routed functions.

File: src/azul/openapi.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        func = next(iter(entries.values())).view_function
+        if hasattr(func, 'api_spec'):
+            docs[route] = func.api_spec
+    return docs
+
+
+def merge_dicts(d1: MutableMapping, d2: MutableMapping, mutate: bool = False, override: bool = False):
+    """
+    Merge two dictionaries deeply such that:
+    - collisions raise an error,
```
Comment (hannes-ucsc) [inline src/azul/openapi.py]:
```suggestion

    - collisions raise an error,
```

File: src/azul/openapi.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    - lists are appended, and
+    - child dictionaries are recursively merged
+    If override is true, d2 overrides d1 on collisions.
```
Comment (hannes-ucsc) [inline src/azul/openapi.py]:
```suggestion

    If override is true, d2 overrides d1 on collisions.
```

Comment (hannes-ucsc) [reply src/azul/openapi.py]: And consider adding empty lines before each case (after all but the last assertion).

File: src/azul/openapi.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    >>> merge_dicts({1: {'a': 'z'}}, {1: {'a': 'y'}}, override=True)
+    {1: {'a': 'y'}}
+    """
+    if mutate:
+        d1 = copy.deepcopy(d1)
+        d2 = copy.deepcopy(d2)
+    return recursive_merge_dicts(d1, d2, [], override)
+
+
+def recursive_merge_dicts(merged: MutableMapping, d2: MutableMapping, path: List[str], override: bool):
```
Comment (hannes-ucsc) [inline src/azul/openapi.py]: If a method is solely for internal use, prefix its name with underscore. It helps communicating what's intended to be used from the outside and what should be considered internal.

File: test/local_integration_test.py [hunk 12]
```diff
@@ -340,6 +342,19 @@ def _get_entities_by_project(self, entity_type, project_short_name):
         return entities
 
 
+class OpenAPIIntegrationTest(unittest.TestCase):
+
+    def test_openapi(self):
+        service = config.service_endpoint()
+        response = requests.get(service + '/')
+        self.assertEqual(response.status_code, 200)
```
Comment (hannes-ucsc) [inline test/local_integration_test.py]: Maybe assert the content type (should be html) and length (should be non-zero).

File: src/azul/openapi.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+    >>> merge_dicts({1: {'a': 'z', 'c': [1]}}, {1: {'b': 'y', 'c': [2]}})
+    {1: {'a': 'z', 'c': [1, 2], 'b': 'y'}}
+    >>> merge_dicts({1: {'a': 'z'}}, {1: {'a': 'y'}})
+    Traceback (most recent call last):
+    ...
+    ValueError: Cannot merge path '1.a' with values: z, y
+    >>> merge_dicts({1: {'a': 'z'}}, {1: {'a': 'y'}}, override=True)
+    {1: {'a': 'y'}}
+    """
+    if mutate:
```
Comment (hannes-ucsc) [inline src/azul/openapi.py]: I don't see this ever being used. If not, just eliminate it. If it is, the naming is odd. mutate=True cause a copy to be made? I'd expect the opposite, mutate=False should make the copy. Also if the function modifies the argument, it should not return anything. So: eliminate if not used or replace this modal argument with a make a wrapper function that copies first.

File: src/azul/openapi.py (hunk 10 above)
Comment (hannes-ucsc) [reply src/azul/openapi.py]: And explain the parameters.

File: src/azul/openapi.py (hunk 13 above)
Comment (hannes-ucsc) [reply src/azul/openapi.py]: BTW, `inplace` is a more common name for this behavior.

File: scripts/apidev.py [hunk 14]
```diff
@@ -0,0 +1,59 @@
+import json
+import os
+import shutil
```
Comment (hannes-ucsc) [inline scripts/apidev.py]: unused

File: scripts/apidev.py (hunk 3 above)
Comment (hannes-ucsc) [reply scripts/apidev.py]:
```diff
Index: scripts/apidev.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- scripts/apidev.py	(revision 7068a20c2848c16d2304c25ffcbd187d239aa565)
+++ scripts/apidev.py	(date 1566351360490)
@@ -1,6 +1,5 @@
 import json
 import os
-import shutil
 from http.server import HTTPServer, SimpleHTTPRequestHandler
 
 from watchdog.events import FileSystemEventHandler
@@ -12,8 +11,6 @@
 
 SPEC_FILE = 'openapi.json'
 parent_dir = os.path.realpath(os.path.dirname(__file__))
-static_dir = os.path.join(parent_dir, 'apidev_static')
-web_page = os.path.join(static_dir, 'swagger-editor.html')
 
 
 def write_specs(gateway_id, app, api_info):
@@ -23,14 +20,17 @@
 
 
 def main():
+    static_dir = os.path.join(parent_dir, 'apidev_static')
+    web_page = os.path.join(static_dir, 'swagger-editor.html')
     service = load_app_module('service')
-
-    server_address = ('localhost', 8787)
-    server_url = f"http://{':'.join(map(str, server_address))}"
-    httpd = HTTPServer(server_address, SimpleHTTPRequestHandler)
+    # That's what made me chuckle earlier today
+    host, port = 'localhost', 8787
+    server_url = f"http://{host}:{port}"
+    httpd = HTTPServer((host, port), SimpleHTTPRequestHandler)
     address = f"{server_url}/{os.path.relpath(web_page)}?url={server_url}/{os.path.relpath(parent_dir)}/{SPEC_FILE}"
     print(f'Open {address} in browser to validate changes.')
-    print('If you changed/added any routes, make sure to deploy!')
+    print('If you changed/added any routes, make sure to `make deploy` afterwards!')
+    # TODO: explain why
 
     gateway_id = aws.api_gateway_id('service', validate=True)
     event_handler = UpdateHandler(service, gateway_id)
```

File: src/azul/deployment.py [hunk 15]
```diff
@@ -88,10 +98,13 @@ def es_endpoint(self) -> Netloc:
 
     @property
     def lambda_env(self) -> Mapping[str, str]:
-        return config.lambda_env(self.es_endpoint)
+        return {
+            **config.lambda_env(self.es_endpoint),
+            'AZUL_SERVICE_API_GATEWAY_ID': aws.api_gateway_id('service', validate=True)
```
[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply src/azul/deployment.py]: Evidence for `inplace`:

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html

Note the return value of None if inplace=True

File: src/azul/openapi.py (hunk 13 above)
Comment (jessebrennan) [reply src/azul/openapi.py]: Woops! The opposite behavior was what was intended. I'll get rid of it then... I don't believe it's covered in doctests. I made it originally because I wasn't sure if I'd need it or not.

File: lambdas/service/lambda-policy.json.template.py (hunk 1 above)
Comment (jessebrennan) [reply lambdas/service/lambda-policy.json.template.py]: Changed.

File: lambdas/service/Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (jessebrennan) [reply lambdas/service/Makefile]: Ok great. I made this change.

File: lambdas/service/app.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jessebrennan) [reply lambdas/service/app.py]: Done.

File: src/azul/deployment.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (jessebrennan) [reply src/azul/deployment.py]: In response to the first comment here, I made the change.

File: scripts/apidev.py (hunk 3 above)
Comment (jessebrennan) [reply scripts/apidev.py]: Ahh... Yeah I had just copied that straight out of the HTTPServer docs. Didn't even think to simplify.

File: lambdas/service/app.py (hunk 6 above)
Comment (jessebrennan) [reply lambdas/service/app.py]: I dig it. The only con I can think of is that it takes some of the OpenAPI logic out of `openapi.py`.

File: src/azul/openapi.py (hunk 8 above)
Comment (jessebrennan) [reply src/azul/openapi.py]: I see. OK I'll give it a whirl.

File: src/azul/deployment.py (hunk 7 above)
Comment (jessebrennan) [reply src/azul/deployment.py]: That was the only one I found. I will more to a separate commit.

File: lambdas/service/app.py (hunk 6 above)
Comment (jessebrennan) [reply lambdas/service/app.py]: I would still prefer the parameter name `spec` over `openapi` however.

File: src/azul/openapi.py (hunk 10 above)
Comment (jessebrennan) [reply src/azul/openapi.py]: Done.

File: src/azul/openapi.py (hunk 11 above)
Comment (jessebrennan) [reply src/azul/openapi.py]: Good point.

File: scripts/apidev.py (hunk 14 above)
Comment (jessebrennan) [reply scripts/apidev.py]: Removed.

File: src/azul/deployment.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply src/azul/deployment.py]: The `inplace` comment was in reference to  https://github.com/DataBiosphere/azul/pull/1227#discussion_r315967458

File: src/azul/chalice.py [hunk 16]
```diff
@@ -13,9 +14,12 @@ def route(self, path, **kwargs):
         Same as method in supper class but stashes URL path a view function is bound to as an attribute of the
         function itself.
         """
+        description = kwargs.pop('spec', None)
```
Comment (hannes-ucsc) [inline src/azul/chalice.py]:
```suggestion
        spec = kwargs.pop('spec', None)
```

File: src/azul/chalice.py [hunk 17]
```diff
@@ -13,9 +14,12 @@ def route(self, path, **kwargs):
...
         decorator = super().route(path, **kwargs)
 
         def _decorator(view_func):
+            if description:
+                spec(description)(view_func)
```
Comment (hannes-ucsc) [inline src/azul/chalice.py]:
```suggestion
                view_func = spec(description)(view_func)
```

This is the decorator protocol. If you didn't assign, you'd be relying on an implementation detail of `spec`.

File: lambdas/service/app.py [hunk 18]
```diff
@@ -43,6 +44,18 @@
 configure_app_logging(app, log)
 
 
+api_info = {
```
Comment (hannes-ucsc) [inline lambdas/service/app.py]: Can we consolidate the terminology a bit? If it is `spec` for the decorator and keyword argument, it should be `spec` here as well. OTOH, `spec` is a bit too generic to be used at the module scope. That's why I'd prefer `openapi` for everything.

File: src/azul/chalice.py (hunk 16 above)
Comment (jessebrennan) [reply src/azul/chalice.py]: This will cause a name collision. I will import the spec decorator so that it's use is qualified (to avoid the collision).

File: src/azul/chalice.py (hunk 17 above)
Comment (jessebrennan) [reply src/azul/chalice.py]: That's what I had originally, but I realized it was redundant. I changed it back however.

File: lambdas/service/app.py (hunk 18 above)
Comment (jessebrennan) [reply lambdas/service/app.py]: What about something like `top_level_spec` or `global_spec`? `openapi` just parses kind of funny for me.

Comment (hannes-ucsc) [reply lambdas/service/app.py]: `top_level_spec` is long and awkward. Both "global" and "toplevel" are implied because the variable is defined at the module scope. I like `openapi` because it is both short and unique. Can you explain what you mean by "parses funny"?

**Comment by hannes-ucsc** [comment]
I am absolutely baffled as to why the builds are failing. One is failing in a health checker test the other one in trufflehog.

**Comment by jessebrennan** [comment]
@hannes-ucsc I could try making a new branch and opening a new PR?

**Comment by hannes-ucsc** [comment]
I'll look into the test_laziness failure because it's happening on other branches, too.

**Comment by hannes-ucsc** [comment]
I did a binary search over the Trufflehog patterns. It' s the Github pattern that triggers this:

https://github.com/DataBiosphere/azul/blob/b82b32b035118a367c63e7fa4d109f298ad539e7/.trufflehog.json#L12

The pattern is a bit odd. In fact, I think it's just dumb. Github API keys are just hex strings of a fixed length so I don't know what that pattern even attempts to match.

**Comment by jessebrennan** [comment]
@hannes-ucsc, I'm not sure what you mean by "without (!) rebasing". I think you mean without rebasing onto the latest develop. I squashed commits properly, but didn't rebase onto develop.
